server/manager/src/main/java/org/apache/accumulo/master/TabletGroupWatcher.java

```diff
@@ -871,7 +870,9 @@ private void flushChanges(SortedMap<TServerInstance,TabletServerStatus> currentT
 
     if (!assignments.isEmpty()) {
       Master.log.info(String.format("Assigning %d tablets", assignments.size()));
-      store.setFutureLocations(assignments);
+
+      for (Assignment assignment : assignments)
+        store.setFutureLocation(assignment);
```

Comment:
It seems like this code would be less efficient, because it creates a new batch writer for each assignment, and that the method that handles the updates in Ample should deal with updating all of the assignments in one pass. What do you know about the performance implications of this change?

Reply:
I am not too sure about the performance implications. I remember this part being one of the ones I struggled with trying to figure out how to apply my other changes too. Part of the original goal of making these changes to the state stores was to pass in a single assignment instead of a collection, along with applying ample to it. Some of the state stores already did this before these changes like in ZooTabletStateStore.setFutureLocation. Its possible the easiest and best solution is to add the collection<assignment> back and use: 
'''java
 if (assignments.size() != 1)
      throw new IllegalArgumentException("There is only one root tablet");
    Assignment assignment = assignments.iterator().next();
    if (assignment.tablet.compareTo(RootTable.EXTENT) != 0)
      throw new IllegalArgumentException("You can only store the root tablet location");
'''
at the beginning of the various state store functions.

Reply:
Well, the reason the ZooTabletStateStore only accepted one was because it can only ever have one... because the root table that uses the ZooTabletStateStore can never have more than one tablet, and therefore, there can never be more than one assignment being updated. That's not true for the other stores, because the metadata table can split into multiple tablets, and so can user tables.

Reply:
That makes sense. I was told that was the case(only one assignment at a time) for all the stores even though we implemented it to be able to handle multiple assignments. Just to be clear, I should revert back to using collection<assignments> for these functions?

Reply:
Based on the code you changed, it does not appear that's the case. Specifically, where you made this update in a loop, it wasn't one at a time. I think you could probably have both the single-case version, and the collection version, so we don't have to do "Collections.singletonList" everywhere when we only have one. The Collection one could be optimized to reuse the same batchwriter somehow (not sure exactly what would need to change in the tablet mutator / Ample code to do that; the old code that didn't use Ample was obviously reusing a batch writer, but I'm not sure how to do it in an Ample way). I'm also not sure how you'd pass the previous location to clear if you're passing a collection... you might have a different previous location for each assignment, right? I'm not sure without analyzing the code in a lot more detail.